src/v1/album.js

```diff
@@ -51,46 +69,173 @@ router.put('/:id(\\w{24})', (req, res) => {
+            Promise.all([artist.save(), album.save()]).then(result => {
+              res.send(new Response.Data(album));
+            });
           }
-          album.artist = artist.toSimple();
-          album.save();
-          res.status(200).send(new Response.Data(album));
         });
+      }
+    });
+  }
+});
+
+router.delete('/:id(\\w{24})/artist', (req, res) => {
+  Album.getAlbum(req.params.id).then(album => {
+    if (!album) {
+      res.status(400)
+        .send(new Response.Error(new Err.ResourceNotFound(['album not found'])));
     } else {
-      res.status(200).send(new Response.Data(album));
+      Artist.getArtist(album.artist._id.toHexString()).then(artist => {
+        _.remove(artist.albums, (_album) => {
+          return _album._id.equals(album._id);
+        });
+        album.artist = null;
+        Promise.all([artist.save(), album.save()]).then(result => {
+          res.send(new Response.Data(album));
+        });
+      });
     }
   });
 });
 
-router.delete('/:id(\\w{24})', (req, res) => {
-  Album.deleteAlbum(req.params.id)
-    .then(result => {
-      res.send(new Response.Data(result));
-    })
-    .catch(err => {
-      if (err instanceof Err.UnremovableError) {
-        res.status(400);
+router.put('/:albumId(\\w{24})/artist', (req, res) => {
+  if (!req.body.artist || req.body.artist == '') {
+    res.status(400)
+      .send(new Response.Error(new Err.InvalidParam(['artist is required'])));
+  }
+
+  Album.getAlbum(req.params.albumId).then(album => {
+    if (!album) {
+      res.status(400)
+        .send(new Response.Error(new Err.ResourceNotFound(['album not found'])));
+    } else {
+      /** Remove album from ori artist. */
+      var oriArtist = null;
+      if (album.artist) {
+        oriArtist = new Promise((resolve, reject) => {
+          Artist.getArtist(album.artist._id.toHexString()).then(artist => {
+            _.remove(artist.albums, (_album) => {
+              return _album._id.equals(album._id);
+            });
+            artist.save().then(result => {
+              resolve();
+            });
+          });
+        });
       }
-      res.send(new Response.Error(err));
-    });
+
+      /** Add album to new artist */
+      var newArtist = null;
+      newArtist = new Promise((resolve, reject) => {
+        Artist.getArtist(req.body.artist).then(artist => {
+          var _album = _.find(artist.albums, (o) => {
+            return o._id.equals(album._id);
+          });
+          if (!_album)
+            artist.albums.push(album.toSimple());
+          album.artist = artist.toSimple();
+          artist.save().then(result => {
+            resolve();
+          });
+        });
+      });
+
+      Promise.all([oriArtist, newArtist]).then(result => {
+        album.save().then(result => {
+          res.send(new Response.Data(album));
+        });
+      });
+    }
+  });
+});
+
+router.post('/:id(\\w{24})/track', (req, res) => {
```

Comment:
I think this api should let the admin adds the array of track to the album.

Reply:
Already added.